c2rust-analyze/src/context.rs

```diff
@@ -1399,3 +1399,127 @@ impl Assignment<'_> {
         )
     }
 }
+
+/// Print an `LTy` as a string, using the provided callback to print the labels on each pointer and
+/// reference type.
+///
+/// Note this completely omits any labels on non-pointer types.
+pub fn print_ty_with_pointer_labels<L: Copy>(
+    lty: LabeledTy<L>,
+    mut f: impl FnMut(L) -> String,
+) -> String {
+    let mut out = String::new();
+    print_ty_with_pointer_labels_into(&mut out, lty, &mut f);
+    out
+}
+
+pub fn print_ty_with_pointer_labels_into<L: Copy>(
+    dest: &mut String,
+    lty: LabeledTy<L>,
+    f: &mut impl FnMut(L) -> String,
+) {
+    use rustc_type_ir::TyKind::*;
+    match lty.ty.kind() {
+        // Types with no arguments
+        Bool | Char | Int(_) | Uint(_) | Float(_) | Str | Foreign(_) | Never => {
+            write!(dest, "{:?}", lty.ty).unwrap();
+        }
+
+        // Types with arguments
+        Adt(adt_def, _substs) => {
+            write!(dest, "{:?}", adt_def.did()).unwrap();
+            if lty.args.len() != 0 {
+                dest.push('<');
+                // TODO: region args
+                for (i, &arg_lty) in lty.args.iter().enumerate() {
+                    if i > 0 {
+                        dest.push_str(", ");
+                    }
+                    print_ty_with_pointer_labels_into(dest, arg_lty, f);
+                }
+                dest.push('>');
+            }
+        }
+        &Array(_elem, len) => {
+            dest.push('[');
+            print_ty_with_pointer_labels_into(dest, lty.args[0], f);
+            write!(dest, "; {:?}]", len).unwrap();
+        }
+        &Slice(_elem) => {
+            dest.push('[');
+            print_ty_with_pointer_labels_into(dest, lty.args[0], f);
+            dest.push(']');
+        }
+        RawPtr(mty) => {
+            if mty.mutbl == Mutability::Not {
+                dest.push_str("*const ");
+            } else {
+                dest.push_str("*mut ");
+            }
+            let s = f(lty.label);
+            if s.len() > 0 {
+                dest.push_str(&s);
+                dest.push_str(" ");
+            }
+            print_ty_with_pointer_labels_into(dest, lty.args[0], f);
+        }
+        &Ref(_rg, _ty, mutbl) => {
+            let s = f(lty.label);
+            if mutbl == Mutability::Not {
+                dest.push_str("&");
+                if s.len() > 0 {
+                    dest.push(' ');
+                }
+            } else {
+                dest.push_str("&mut ");
+            }
+            if s.len() > 0 {
+                dest.push_str(&s);
+                dest.push_str(" ");
+            }
+            print_ty_with_pointer_labels_into(dest, lty.args[0], f);
+        }
+        FnDef(def_id, _substs) => {
+            write!(dest, "{:?}", def_id).unwrap();
```

Comment:
I don't remember what's in the `DefId`, does this print a function name?

Reply:
Handily we've written a parser for this format in `parse_def_id` - the name is included if available.

Reply:
> I don't remember what's in the `DefId`, does this print a function name?

It will print something like `DefId(0:4 ~ insertion_sort[add9]::insertion_sort)`, which is not the prettiest, but should be okay for debug output.